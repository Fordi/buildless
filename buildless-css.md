`buildless-css`
---------------

This is a minimalist CSS-in-JS tagged template that seeks to provide the same
benefit as CSS modules in a webpack/node build environment, but lets you work
on a component's CSS co-located with the component, and running natively
in-browser.  The general form is as follows:

```javascript
import { html, css } from 'https://unpkg.com/@fordi-org/buildless';
import { colors } from './theme.js'
const styles = css`
  .about {
    background: ${colors.background};
    color: ${colors.text};
  }
`;

export default ({ isBlue }) => html`
  <article className=${styles.about.and('some-other-style', isBlue && 'about--blue')}>
    <h1>About</h1>
    <p>
      Neat stuff goes here
    </p>
  </article>
`;
```

This will work uncompiled, but also, the production build will automagically
minify any CSS used here.

What `css` returns is an object containing decorated class names.  Technically,
they're objects with a `toString()` function, plus an `and(...classList)`
member, which accepts a list of classNames and falsy values (which are skipped).

If you need to write a list of regular (non-parsed) class names (for example, to
add to a `className` prop, or to reference global styles), you can use the
export `classes(...classList)`, which works in exactly the same way, e.g.,

```javascript
  export default ({ isFeatured, children }) => html`
    <article className=${classes('main', isFeatured && 'main--featured')}>
      ${children}
    </article>
  `;
```

Both `and()` and `classes()` return decorated class name objects, so you can
chain away to your heart's content.
