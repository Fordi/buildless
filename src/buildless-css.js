const joinClass = (...classes) => Array.from(new Set(classes.filter(a => !!a).join(' ').split(' ').reverse())).reverse().join(' ');

export const classes = (...names) => Object.assign(
  (...more) => classes(...names, ...more),
  {
    toString: () => joinClass(...names),
    and: (...more) => classes(...names, ...more),
  }
);

const createStylesheet = (str) => {
  const rid = Math.random().toString(36).substr(2);
  const cssRules = parseRules(str);
  const styles = {};
  cssRules.forEach(rule => Object.assign(styles, insertRule(rule, rid)));
  Object.keys(styles).forEach(name => {
    styles[name] = classes(styles[name])
  });
  return styles;
};

const head = document.querySelector('head');
const { sheet } = head.appendChild(document.createElement('style'));

const clsRx = /\.([^ \.\[:>,]+)/g;
export const css = (...args) => createStylesheet(String.raw(...args));

const appendId = (str, id) => str.endsWith(`_${id}`) ? str : `${str}_${id}`;
const aniKwRx = /infinite|none|forwards|backwards|both|paused|running|normal|reverse|alternate-normal|alternate-reverse/;

const insertRule = (rule, id) => {
  const classNames = allRules(rule).reduce((cls, r) => {
    if (r.type === CSSRule.KEYFRAMES_RULE) {
      r.name = appendId(r.name, id);
    }
    if (r.style) {
      if (r.style['animation-name']) {
        r.style['animation-name'] = r.style['animation-name'].split(',').map(p => appendId(p.trim(), id)).join(',');
      }
      if (r.style.animation) {
        r.style.animation = r.style.animation.replace(/,[\s\r\n\t]+/g, ',').split(/[\s\r\n\t]+/).map(part => {
          // Can have multiple values
          const [piece, ...pieces] = part.split(',');
          // Starts with a number; it's a time value or iteration count
          if (/^[\.0-9]/.test(piece)) return part;
          // keywords
          if (aniKwRx.test(piece)) return part;
          // Not a keyword or numeric; it's an ID.
          return [piece, ...pieces].map(p => appendId(p.trim(), id)).join(',');
        }).join(' ');
      }
    }
    if (r.selectorText) {
      r.selectorText = r.selectorText.replace(clsRx, (_, m) => {
        cls[m] = appendId(m, id);
        return `.${cls[m]}`;
      });
    }
    return cls;
  }, {});
  sheet.insertRule(rule.cssText);
  return classNames;
};

const allRules = a => (
  a.selectorText
    ? [a]
    : Array.from(a.cssRules || []).reduce((list, rule) => (
      [a, ...list, ...allRules(rule)]
    ), [])
);

const parseRules = css => {
  const t = document.createElement('style');
  t.textContent = css;
  head.appendChild(t);
  const r = t.sheet;
  head.removeChild(t);
  return Array.from(r.cssRules);
};
