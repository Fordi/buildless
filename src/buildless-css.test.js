import { classes, css } from './buildless-css.js';

describe('classes', () => {
  it('creates a filtered class list with early duplicates folded', () => {
    expect(classes('one', false && 'two', 'three', 'one').toString()).toBe('three one')
  });
  it('is chainable', () => {
    expect(classes('one')(false && two).and('three')('one').toString()).toBe('three one');
  });
});
const localMath = global.Math;
describe('css', () => {
  beforeEach(() => {
    const mockMath = Object.create(global.Math);
    // ensures that Math.random().toString(36) always returns 0.deadbeef
    mockMath.random = () => 0.3721358433517089;
    global.Math = mockMath;
  });
  afterEach(() => {
    for (const sheet of document.querySelectorAll('head>style')) {
      sheet.parentNode.removeChild(sheet);
    };
    global.Math = localMath;
  });
  it('returns a hash of class names', () => {
    const styles = css`
      .one {
        padding: 0;
      }
      .two {
        margin: 0;
        animation-name: snarf;
      }
      .three {
        animation: snarf 5s linear 2s infinite alternate;
      }
      @keyframes snarf {
        0% { border: 1px solid black; }
        100% { border: 100px solid white; }
      }
    `;
    const sheets = document.querySelectorAll('style');
    expect(sheets.length).toBe(1);
    ['one', 'two', 'three'].forEach(name => {
      expect(styles[name]).toBeDefined();
      expect(styles[name].toString()).not.toBe(name);
      expect(styles[name].toString()).toBe(`${name}_deadbeef`);
    });
    expect(sheets[0].sheet).toMatchSnapshot();
  });
});
