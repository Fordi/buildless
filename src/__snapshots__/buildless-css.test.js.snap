// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css returns a hash of class names 1`] = `
CSSStyleSheet {
  "cssRules": Array [
    CSSKeyframesRule {
      "__ends": 100,
      "__starts": 0,
      "_vendorPrefix": undefined,
      "cssRules": Array [
        CSSKeyframeRule {
          "__ends": 60,
          "__starts": 34,
          "keyText": "0%",
          "parentRule": [Circular],
          "parentStyleSheet": CSSStyleSheet {
            "cssRules": Array [
              [Circular],
            ],
            "parentStyleSheet": null,
          },
          "style": CSSStyleDeclaration {
            "0": "border",
            "_importants": Object {
              "border": "",
            },
            "border": "1px solid black",
            "length": 1,
            "parentRule": [Circular],
          },
        },
        CSSKeyframeRule {
          "__ends": 97,
          "__starts": 69,
          "keyText": "100%",
          "parentRule": [Circular],
          "parentStyleSheet": CSSStyleSheet {
            "cssRules": Array [
              [Circular],
            ],
            "parentStyleSheet": null,
          },
          "style": CSSStyleDeclaration {
            "0": "border",
            "_importants": Object {
              "border": "",
            },
            "border": "100px solid white",
            "length": 1,
            "parentRule": [Circular],
          },
        },
      ],
      "name": "snarf_deadbeef",
      "parentRule": null,
      "parentStyleSheet": [Circular],
    },
    CSSStyleRule {
      "__ends": 94,
      "__starts": 0,
      "parentRule": null,
      "parentStyleSheet": [Circular],
      "selectorText": ".three_deadbeef",
      "style": CSSStyleDeclaration {
        "0": "animation",
        "__starts": 16,
        "_importants": Object {
          "animation": "",
        },
        "animation": "snarf_deadbeef 5s linear_deadbeef 2s infinite alternate_deadbeef",
        "length": 1,
        "parentRule": [Circular],
      },
    },
    CSSStyleRule {
      "__ends": 58,
      "__starts": 0,
      "parentRule": null,
      "parentStyleSheet": [Circular],
      "selectorText": ".two_deadbeef",
      "style": CSSStyleDeclaration {
        "0": "margin",
        "1": "animation-name",
        "__starts": 14,
        "_importants": Object {
          "animation-name": "",
          "margin": "",
        },
        "animation-name": "snarf_deadbeef",
        "length": 2,
        "margin": "0",
        "parentRule": [Circular],
      },
    },
    CSSStyleRule {
      "__ends": 27,
      "__starts": 0,
      "parentRule": null,
      "parentStyleSheet": [Circular],
      "selectorText": ".one_deadbeef",
      "style": CSSStyleDeclaration {
        "0": "padding",
        "__starts": 14,
        "_importants": Object {
          "padding": "",
        },
        "length": 1,
        "padding": "0",
        "parentRule": [Circular],
      },
    },
  ],
  "parentStyleSheet": null,
}
`;
